rts/motoko-rts/src/gc/generational.rs

```diff
@@ -0,0 +1,499 @@
+    };
+    let heap = Heap {
+        mem,
+        limits: get_limits(),
+        roots,
+    };
+    let strategy = decide_strategy(&heap.limits);
+    #[cfg(debug_assertions)]
+    let forced_gc = strategy.is_none();
+    let strategy = strategy.unwrap_or(Strategy::Young);
+    let mut gc = GenerationalGC::new(heap, strategy);
+
+    #[cfg(debug_assertions)]
+    sanity_checks::verify_snapshot(&gc.heap, false);
+
+    gc.run();
+
+    let new_limits = &gc.heap.limits;
+    set_limits(&gc.heap.limits);
+    update_statistics(&old_limits, new_limits);
+    update_strategy(strategy, new_limits);
+
+    #[cfg(debug_assertions)]
+    if !forced_gc {
+        sanity_checks::check_memory(&gc.heap.limits, &gc.heap.roots);
+        sanity_checks::take_snapshot(&mut gc.heap);
+    }
+
+    write_barrier::init_write_barrier(gc.heap.mem);
+}
+
+#[cfg(feature = "ic")]
+unsafe fn get_limits() -> Limits {
+    use crate::memory::ic;
+    Limits {
+        base: ic::get_aligned_heap_base() as usize,
+        last_free: core::cmp::max(ic::LAST_HP as usize, ic::get_aligned_heap_base() as usize), // max because of aligned heap base
+        free: ic::HP as usize,
+    }
+}
+
+#[cfg(feature = "ic")]
+unsafe fn set_limits(limits: &Limits) {
+    use crate::memory::ic;
+    ic::HP = limits.free as u32;
+    ic::LAST_HP = limits.free as u32;
+}
+
+#[cfg(feature = "ic")]
+unsafe fn update_statistics(old_limits: &Limits, new_limits: &Limits) {
+    use crate::memory::ic;
+    let live_size = Bytes(new_limits.free as u32 - new_limits.base as u32);
+    ic::MAX_LIVE = ::core::cmp::max(ic::MAX_LIVE, live_size);
+    ic::RECLAIMED += Bytes(old_limits.free as u64 - new_limits.free as u64);
+}
+
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum Strategy {
+    Young,
+    Full,
+}
+
+#[cfg(feature = "ic")]
+static mut OLD_GENERATION_THRESHOLD: usize = 32 * 1024 * 1024;
+
+#[cfg(feature = "ic")]
+static mut PASSED_CRITICAL_LIMIT: bool = false;
+
+#[cfg(feature = "ic")]
+const CRITICAL_MEMORY_LIMIT: usize = (4096 - 512) * 1024 * 1024;
+
+#[cfg(feature = "ic")]
+unsafe fn decide_strategy(limits: &Limits) -> Option<Strategy> {
+    const YOUNG_GENERATION_THRESHOLD: usize = 8 * 1024 * 1024;
+
+    let old_generation_size = limits.last_free - limits.base;
+    let young_generation_size = limits.free - limits.last_free;
+
+    if limits.free >= CRITICAL_MEMORY_LIMIT && !PASSED_CRITICAL_LIMIT {
+        PASSED_CRITICAL_LIMIT = true;
+        Some(Strategy::Full)
+    } else if old_generation_size > OLD_GENERATION_THRESHOLD {
+        Some(Strategy::Full)
+    } else if young_generation_size > YOUNG_GENERATION_THRESHOLD {
+        Some(Strategy::Young)
+    } else {
+        None
+    }
+}
+
+#[cfg(feature = "ic")]
+unsafe fn update_strategy(strategy: Strategy, limits: &Limits) {
+    const GROWTH_RATE: f64 = 1.5;
+    if strategy == Strategy::Full {
+        OLD_GENERATION_THRESHOLD = (limits.free as f64 * GROWTH_RATE) as usize;
```

Comment:
```suggestion
        OLD_GENERATION_THRESHOLD = (limits.free - limits.base) as f64 * GROWTH_RATE) as usize;
```
It looks like this is a threshold on  ``old_generation_size`, given the comparison in `decide_strategy` on line 127. So I guess you should be subtracting limits.base (or use a common helper)

Reply:
Indeed. I fixed this. Thank you.